File: backend/domain/learning/learning_interaction.py
from __future__ import annotations

from datetime import UTC, datetime
from uuid import uuid4

from pydantic import BaseModel, Field


class LearningInteraction(BaseModel):
    """
    Represents one observable learning interaction.

    An interaction is evidence used to update the
    learner model.
    """

    id: str = Field(
        default_factory=lambda: str(uuid4()),
    )

    learner_id: str

    question_id: str | None = None

    question: str = ""

    answer: str = ""

    correct: bool | None = None

    concept_ids: list[str] = Field(
        default_factory=list,
    )

    timestamp: datetime = Field(
        default_factory=lambda: datetime.now(UTC),
    )

    metadata: dict[str, object] = Field(
        default_factory=dict,
    )

File: frontend/lib/domain/models/learning/learning_interaction.dart
class LearningInteraction {
  final String id;

  final String learnerId;

  final String? questionId;

  final String question;

  final String answer;

  final bool? correct;

  final List<String> conceptIds;

  final DateTime timestamp;

  final Map<String, dynamic> metadata;

  const LearningInteraction({
    required this.id,
    required this.learnerId,
    this.questionId,
    this.question = '',
    this.answer = '',
    this.correct,
    this.conceptIds = const [],
    required this.timestamp,
    this.metadata = const {},
  });
}
